keps/sig-auth/apiserver_auth_to_webhooks.md

```diff
@@ -0,0 +1,128 @@
+      * [Risks and Mitigations](#risks-and-mitigations)
+  * [Graduation Criteria](#graduation-criteria)
+  * [Implementation History](#implementation-history)
+  * [Alternatives](#alternatives)
+
+## Summary
+
+We want to provide a simple means of authenticating outgoing webhooks from the apiserver and its aggregates.
+
+## Motivation
+
+Outgoing webhooks from the apiserver such as the ones found in [dynamic admission control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers) and [dynamic audit](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#dynamic-backend) 
+suffer from a lack of easily configurable authentication. Currently, Dynamic Admission webhooks provide a mechanism for plugin authentication by a kubeconfig provisioned on the host. The intention of this KEP is to provide a simpler means for the receiving 
+server to authenticate the apiserver and its aggregates using the [Authentication API](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go).
+
+### Goals
+* A simple means of authenticating apiserver clients.
+
+### Non-Goals
+* Providing all the authentication schemes found in the current kubeconfig.
+
+## Proposal
+
+We propose a simple mechanism for authenticating webhooks using the token [Authentication API](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go). The shared 
+[webhook client](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/webhook/client.go) will be parameterized to optionally enhance every outgoing request with a token obtained from a  
+[Token Request](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go#L112). The receiving server can then check that token using a [Token Review](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go#L45).
+
+### User Stories
+
+#### Story 1
+I am a cluster administrator using the dynamic auditing feature and want to make sure the logs I receive are from the apiserver.
+
+#### Story 2
+I am a plugin developer and want to easily authenticate the apiserver.
+
+### Implementation Details/Notes/Constraints
+
+A new struct will be added to the [client config](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/webhook/client.go#L40) for outgoing webhooks:
+
+```go
+type AuthInfo struct {
+  ProvisionToken  bool
+}
+
+type ClientConfig struct {
+  Name     string
+  URL      string
+  CABundle []byte
+  AuthInfo *AuthInfo
+  Service  *ClientConfigService
+}
+```
+If enabled the client will provision a token and enhance the outgoing request with an auth header:
+```
+Authorization: bearer <token>
+```
+
+The client will check and refresh the token when necessary. The server can now check the token using a 
+[TokenReview](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go#L45).
```

Comment:
Would be nice to also allow for the tokens to use proper OIDC, if the K8S server is integrated with such a IDP.

The audience-based tokens mounted in Pods seem to also support OIDC in some cases, would be great to be consistent.

Reply:
I agree with this suggestion.

This can also solve the case where the webhook/audit-endpoint is not kubernetes-aware. From a quick glance, this seems to be a problem that can eventually be solved by Istio and SPIFFE. I'm guessing the author is trying to provide a better user experience that can quickly establish auth with a simple flag/config.